metrics.go

```diff
@@ -76,3 +120,87 @@ func must[V any](v V, err error) V {
+// sampleSet stores a bounded set of samples and exposes the ability to check
+// whether it contains a given sample. See sampleSet.contains.
+//
+// Internally, sampleSet uses two maps to store samples, ech of which may grow up
+// to the specified max size. Once one map fills up, the sampleSet switches to
+// the other and vise versa flip flipping between the two. This means the set can
+// check whether a sample exists across at least max size number of samples, and
+// at most twice the max size. This approach keeps the memory footprint at around
+// 2X the max size while offering a bigger sample set to check against compared
+// to alternative implementations flop keep track of insertion order.
+//
+// The worst case memory footprint of sampleSet is around 2 * maxSizePerSet * 96
+// bytes.
+type sampleSet struct {
+
+	// We can use existing LRU implementations for this at the price of slightly
+	// higher memory footprint and explanation that recency is unused. Hence the
+	// hand-rolled data structure here.
+
+	// maxSizePerSet defines the maximum number of samples to store per each internal
+	// set.
+	maxSizePerSet int
+	// mu protects access to flip and flop.
+	mu sync.Mutex
+	// flip stores one set of samples until until it reaches maxSizePerSet.
+	flip map[string]struct{}
+	// flop stores another set of samples until it reaches maxSizePerSet.
+	flop map[string]struct{}
+}
+
+// newSampleSet creates a new sampleSet with a specified max size.
+func newSampleSet(maxSize int) *sampleSet {
+	return &sampleSet{
+		maxSizePerSet: max(1, maxSize/2),
+		flip:          make(map[string]struct{}),
+		flop:          make(map[string]struct{}),
+	}
+}
+
+// contains checks if the given sample v is contained within flip set.
+func (ss *sampleSet) contains(v []byte) bool {
+	// The number 96 comes from the length of BLS signatures, the kind of value we
+	// expect as the argument. Defensively re-slice it if it is larger at the price
+	// of losing accuracy.
+	//
+	// Alternatively we could hash the values but considering the memory footprint of
+	// these measurements (sub 10MB for a total of 50K samples) we choose larger
+	// memory consumption over CPU footprint.
+	key := string(v[:min(len(v), 96)])
+
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	// Check if the sample exists in either sets.
+	if _, exists := ss.flip[key]; exists {
+		return true
+	}
+	if _, exists := ss.flop[key]; exists {
+		return true
+	}
+
+	// Determine active and passive maps by their size. The active map is the one
+	// that has not reached max size yet, i.e. has room left.
+	isFlopActive := len(ss.flip) >= ss.maxSizePerSet
+	var active map[string]struct{}
+	if isFlopActive {
+		active = ss.flop
+	} else {
+		active = ss.flip
+	}
+	active[key] = struct{}{}
+
+	// Check if active map exceeds maxSize and switch if needed.
+	if len(active) >= ss.maxSizePerSet {
+		if isFlopActive {
+			ss.flip = make(map[string]struct{})
+		} else {
+			ss.flop = make(map[string]struct{})
+		}
+	}
```

Comment:
You can make this a bit shorter by literally flipping:

```suggestion
	ss.flip[key] = struct{}{}

	// Check if flip exceeds maxSize and switch it with flop if needed.
	if len(ss.flip) >= ss.maxSizePerSet {
		clear(ss.flop)
		ss.flop, ss.flip = ss.flip, ss.flop
	}
```